Assets/SnakeManager.cs

```diff
@@ -0,0 +1,181 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using UnityEngine.UI;
+using Random = UnityEngine.Random;
+
+public class SnakeManager : MonoBehaviour
+{
+    // Start is called before the first frame update
+    public float speed = 1f;
+    
+    public Vector2Int food;
+    public List<Vector2Int> snake;
+    
+    public List<Vector2Int> obstacles;
+    
+    
+    public Vector3 foodPosition;
+    public List<GameObject> snakeDisplay;
+
+    public Vector2Int direction = new Vector2Int(1, 0);
+    public Vector2Int tempDirection = new Vector2Int(1,0);
+    
+    
+    
+    public float time;
+    
+
+    [SerializeField] private GameObject head;
+    [SerializeField] private GameObject foodObject;
+    [SerializeField] private GameObject bodyPrefab;
+    [SerializeField] private Text scoreText;
+    [SerializeField] private GameObject replayButton;
+    
+    public int score;
+    
+    void Start()
+    {
+        time = 0f;
+        score = 0;
+        food = new Vector2Int(4, 0);
+        snake = new List<Vector2Int>
+        {
+            new Vector2Int(0, 0),
+            new Vector2Int(0, 0)
+        };
+        snakeDisplay = new List<GameObject>
+        {
+            head,
+            Instantiate(bodyPrefab)
+        };
+        //create obstacles
+        obstacles = new List<Vector2Int>();
+        int[] values = { -1, 1 };
+        foreach (int i in values)
+        {
+            foreach (int j in values)
+            {
+                obstacles.Add(new Vector2Int(7 * i, 7 * j));
+                obstacles.Add(new Vector2Int(7 * i, 6 * j));
+                obstacles.Add(new Vector2Int(6 * i, 7 * j));
+            }
+        }
+
+        
+
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        time += Time.deltaTime;
+        if (Input.GetKeyDown(KeyCode.UpArrow))
+        {
+            tempDirection = new Vector2Int(0, 1);
+        }
+        else if (Input.GetKeyDown(KeyCode.DownArrow))
+        {
+            tempDirection = new Vector2Int(0, -1);
+        }
+        else if (Input.GetKeyDown(KeyCode.LeftArrow))
+        {
+            tempDirection = new Vector2Int(-1, 0);
+        }
+        else if (Input.GetKeyDown(KeyCode.RightArrow))
+        {
+            tempDirection = new Vector2Int(1, 0);
+        }
+        
+        if (time*speed> 1f)
+        {
+            time -= 1f/speed;
+            Vector2Int newHead = snake[0] + direction;
+            snake.RemoveAt(0);
+            snake.Insert(0, newHead);
+            snake.Insert(0, newHead);
+            snake.RemoveAt(snake.Count-1);
```

Comment:
i dont understand whats going on

remvoed the first element, add newhead to the beginning twice, and remove the last?

Reply:
why 2 heads

Reply:
<img width="509" alt="Screenshot 2024-04-03 at 5 23 12 PM" src="https://github.com/Ad00re/SnakeGame/assets/37167982/c9a7550c-54fc-46f2-b84d-a330969eca1a">
<img width="370" alt="Screenshot 2024-04-03 at 5 27 58 PM" src="https://github.com/Ad00re/SnakeGame/assets/37167982/02b5a78c-759a-4f0e-80f4-ddfea13a8f56">

Since when the snake turns, I want the corner to look like the first one instead of the second one, thus I keep the position of the body that are in middle of the snake change only when the model's value change, and the head and tail lerp depending on the time,  without having 2 head, there will be a gap between the head and the rest of the snake's body.

Reply:
thats redundant. can you modify the program so that each segment corresponds to a snake body strictly? you can store the head position separately.

Reply:
modified